README.md
Package mgmterror provides the standard NETCONF and YANG errors as
defined in RFC6241 and RFC6020 respectively.


debian/changelog
golang-github-danos-mgmterror (1:1.8) unstable; urgency=medium

  [ William Ivory ]
  * Fixups for Bullseye / Debian11 build

 -- Paul Carson <paul.carson@intl.att.com>  Thu, 01 Jul 2021 13:57:10 +0100

golang-github-danos-mgmterror (1:1.7) unstable; urgency=medium

  * Combine and export errpath functions to ErrPath
  * Add InvalidPrefixError

 -- William Ivory <william.ivory@intl.att.com>  Tue, 23 Feb 2021 13:46:05 +0000

golang-github-danos-mgmterror (1:1.6) unstable; urgency=medium

  * Add support for properly checking InfoTags in ExpMgmtError

 -- William Ivory <william.ivory@intl.att.com>  Thu, 10 Dec 2020 16:35:54 +0000

golang-github-danos-mgmterror (1:1.5) unstable; urgency=medium

  * Add AppTag() to Formattable interface
  * Add IsMgmtErrorRef interface
  * Add mgmterrorlist accessor
  * Extend ExpMgmtError.

 -- William Ivory <william.ivory@intl.att.com>  Fri, 25 Sep 2020 08:47:21 +0100

golang-github-danos-mgmterror (1:1.4) unstable; urgency=medium

  * DANOS Import Master

 -- Vyatta Package Maintainers <DL-vyatta-help@att.com>  Fri, 15 Nov 2019 15:40:54 +0000

golang-github-danos-mgmterror (1:1.3) unstable; urgency=medium

  * DANOS Import

 -- Vyatta Package Maintainers <DL-vyatta-help@att.com>  Tue, 12 Nov 2019 10:54:56 +0000


yerror_test.go
// Copyright (c) 2017,2019, AT&T Intellectual Property.  All rights reserved.
//
// Copyright (c) 2016-2017 by Brocade Communications Systems, Inc.
// All rights reserved.
//
// SPDX-License-Identifier: MPL-2.0

package mgmterror

import (
	"bytes"
	"encoding/json"
	"fmt"
	"html"
	"testing"
)

const error_type = "application"

func genNonUniqueXml(basepath string, paths []string) string {
	var error_info bytes.Buffer
	for _, path := range paths {
		error_info.WriteString("\n\t\t<")
		error_info.WriteString(non_unique_info.String())
		error_info.WriteString(" xmlns=\"")
		error_info.WriteString(yang_namespace)
		error_info.WriteString("\">")
		error_info.WriteString(html.EscapeString(path))
		error_info.WriteString("</")
		error_info.WriteString(non_unique_info.String())
		error_info.WriteString(">")
	}
	return `<rpc-error xmlns="` + netconf_namespace + `">
	<error-type>` + html.EscapeString(error_type) + `</error-type>
	<error-tag>` + html.EscapeString(operation_failed.String()) + `</error-tag>
	<error-severity>` + html.EscapeString(yang_severity_error.String()) + `</error-severity>
	<error-app-tag>` + html.EscapeString(data_not_unique.String()) + `</error-app-tag>
	<error-path>` + html.EscapeString(basepath) + `</error-path>
	<error-message>` + html.EscapeString(msg_yang_operation_failed) + `</error-message>
	<error-info>` + error_info.String() + `
	</error-info>
</rpc-error>`
}

func TestNonUniqueError(t *testing.T) {
	basepath := "/testcontainer/testlist"
	paths := []string{
		basepath + "/foo/bar/baz",
		basepath + "/foo/biz/baz",
	}
	ncerr := NewNonUniqueError(paths)
	ncerr.Path = basepath
	marshal, err := json.MarshalIndent(ncerr, "", "\t")
	if err != nil {
		t.Errorf("Marshal NonUniqueError error: %v\n", err)
		return
	}
	unmarshal := NonUniqueError{}
	if err := json.Unmarshal(marshal, &unmarshal); err != nil {
		t.Errorf("Unmarshal NonUniqueError error: %v\n", err)
		return
	}
	cmpMgmtError(t, ncerr.MgmtError, unmarshal.MgmtError)

	verifyXmlMarshal(t, ncerr, genNonUniqueXml(basepath, paths))
}

func ExampleNonUniqueError() {
	basepath := "/testcontainer/testlist"
	paths := []string{
		basepath + "/name/dev1/attr/value",
		basepath + "/name/dev2/attr/value",
		basepath + "/name/dev3/attr/value",
	}
	err := NewNonUniqueError(paths)
	err.Path = basepath
	fmt.Println(err.Error())

	// Output:
	// Error: /testcontainer/testlist: Non-unique paths name/dev1/attr/value, name/dev2/attr/value, name/dev3/attr/value
}

func genTooManyElementsXml(path string) string {
	return `<rpc-error xmlns="` + netconf_namespace + `">
	<error-type>` + html.EscapeString(error_type) + `</error-type>
	<error-tag>` + html.EscapeString(operation_failed.String()) + `</error-tag>
	<error-severity>` + html.EscapeString(yang_severity_error.String()) + `</error-severity>
	<error-app-tag>` + html.EscapeString(too_many_elements.String()) + `</error-app-tag>
	<error-path>` + html.EscapeString(path) + `</error-path>
	<error-message>` + html.EscapeString(msg_yang_operation_failed) + `</error-message>
</rpc-error>`
}

func TestTooManyElementsError(t *testing.T) {
	const path = "/foo/bar/baz"
	ncerr := NewTooManyElementsError(path)
	marshal, err := json.MarshalIndent(ncerr, "", "\t")
	if err != nil {
		t.Errorf("Marshal TooManyElementsError error: %v\n", err)
		return
	}
	unmarshal := TooManyElementsError{}
	if err := json.Unmarshal(marshal, &unmarshal); err != nil {
		t.Errorf("Unmarshal TooManyElementsError error: %v\n", err)
		return
	}
	cmpMgmtError(t, ncerr.MgmtError, unmarshal.MgmtError)

	verifyXmlMarshal(t, ncerr, genTooManyElementsXml(path))
}

func ExampleTooManyElementsError() {
	err := NewTooManyElementsError("biz")
	err.Path = "/foo/bar"
	fmt.Println(err.Error())

	// Output:
	// Error: /foo/bar: The requested operation failed.
}

func genTooFewElementsXml(path string) string {
	return `<rpc-error xmlns="` + netconf_namespace + `">
	<error-type>` + html.EscapeString(error_type) + `</error-type>
	<error-tag>` + html.EscapeString(operation_failed.String()) + `</error-tag>
	<error-severity>` + html.EscapeString(yang_severity_error.String()) + `</error-severity>
	<error-app-tag>` + html.EscapeString(too_few_elements.String()) + `</error-app-tag>
	<error-path>` + html.EscapeString(path) + `</error-path>
	<error-message>` + html.EscapeString(msg_yang_operation_failed) + `</error-message>
</rpc-error>`
}

func TestTooFewElementsError(t *testing.T) {
	const path = "/foo/bar/baz"
	ncerr := NewTooFewElementsError(path)
	marshal, err := json.MarshalIndent(ncerr, "", "\t")
	if err != nil {
		t.Errorf("Marshal TooFewElementsError error: %v\n", err)
		return
	}
	unmarshal := TooFewElementsError{}
	if err := json.Unmarshal(marshal, &unmarshal); err != nil {
		t.Errorf("Unmarshal TooFewElementsError error: %v\n", err)
		return
	}
	cmpMgmtError(t, ncerr.MgmtError, unmarshal.MgmtError)

	verifyXmlMarshal(t, ncerr, genTooFewElementsXml(path))
}

func genMustViolationXml(path string) string {
	return `<rpc-error xmlns="` + netconf_namespace + `">
	<error-type>` + html.EscapeString(error_type) + `</error-type>
	<error-tag>` + html.EscapeString(operation_failed.String()) + `</error-tag>
	<error-severity>` + html.EscapeString(yang_severity_error.String()) + `</error-severity>
	<error-app-tag>` + html.EscapeString(must_violation.String()) + `</error-app-tag>
	<error-path>` + html.EscapeString(path) + `</error-path>
	<error-message>` + html.EscapeString(msg_yang_operation_failed) + `</error-message>
</rpc-error>`
}

func TestMustViolationError(t *testing.T) {
	const path = "/foo/bar/baz"
	ncerr := NewMustViolationError()
	ncerr.Path = path
	marshal, err := json.MarshalIndent(ncerr, "", "\t")
	if err != nil {
		t.Errorf("Marshal MustViolationError error: %v\n", err)
		return
	}
	unmarshal := MustViolationError{}
	if err := json.Unmarshal(marshal, &unmarshal); err != nil {
		t.Errorf("Unmarshal MustViolationError error: %v\n", err)
		return
	}
	cmpMgmtError(t, ncerr.MgmtError, unmarshal.MgmtError)

	verifyXmlMarshal(t, ncerr, genMustViolationXml(path))
}

func genInstanceRequiredXml(path string) string {
	return `<rpc-error xmlns="` + netconf_namespace + `">
	<error-type>` + html.EscapeString(error_type) + `</error-type>
	<error-tag>` + html.EscapeString(data_missing.String()) + `</error-tag>
	<error-severity>` + html.EscapeString(yang_severity_error.String()) + `</error-severity>
	<error-app-tag>` + html.EscapeString(instance_required.String()) + `</error-app-tag>
	<error-path>` + html.EscapeString(path) + `</error-path>
	<error-message>` + html.EscapeString(msg_yang_data_missing) + `</error-message>
</rpc-error>`
}

func TestInstanceRequiredError(t *testing.T) {
	const path = "/foo/bar/baz"
	ncerr := NewInstanceRequiredError(path)
	marshal, err := json.MarshalIndent(ncerr, "", "\t")
	if err != nil {
		t.Errorf("Marshal InstanceRequiredError error: %v\n", err)
		return
	}
	unmarshal := InstanceRequiredError{}
	if err := json.Unmarshal(marshal, &unmarshal); err != nil {
		t.Errorf("Unmarshal InstanceRequiredError error: %v\n", err)
		return
	}
	cmpMgmtError(t, ncerr.MgmtError, unmarshal.MgmtError)

	verifyXmlMarshal(t, ncerr, genInstanceRequiredXml(path))
}

func genLeafrefMismatchXml(path string) string {
	return `<rpc-error xmlns="` + netconf_namespace + `">
	<error-type>` + html.EscapeString(error_type) + `</error-type>
	<error-tag>` + html.EscapeString(data_missing.String()) + `</error-tag>
	<error-severity>` + html.EscapeString(yang_severity_error.String()) + `</error-severity>
	<error-app-tag>` + html.EscapeString(instance_required.String()) + `</error-app-tag>
	<error-path>` + html.EscapeString(path) + `</error-path>
	<error-message>` + html.EscapeString(msg_yang_data_missing) + `</error-message>
</rpc-error>`
}

func TestLeafrefMismatchError(t *testing.T) {
	const path = "/foo/bar/baz"
	ncerr := NewLeafrefMismatchError(path, "foo")
	marshal, err := json.MarshalIndent(ncerr, "", "\t")
	if err != nil {
		t.Errorf("Marshal LeafrefMismatchError error: %v\n", err)
		return
	}
	unmarshal := LeafrefMismatchError{}
	if err := json.Unmarshal(marshal, &unmarshal); err != nil {
		t.Errorf("Unmarshal LeafrefMismatchError error: %v\n", err)
		return
	}
	cmpMgmtError(t, ncerr.MgmtError, unmarshal.MgmtError)

	verifyXmlMarshal(t, ncerr, genLeafrefMismatchXml(path))
}

func genMissingChoiceXml(path, name string) string {
	return `<rpc-error xmlns="` + netconf_namespace + `">
	<error-type>` + html.EscapeString(error_type) + `</error-type>
	<error-tag>` + (operation_failed.String()) + `</error-tag>
	<error-severity>` + html.EscapeString(yang_severity_error.String()) + `</error-severity>
	<error-app-tag>` + html.EscapeString(missing_choice.String()) + `</error-app-tag>
	<error-path>` + html.EscapeString(path) + `</error-path>
	<error-message>` + html.EscapeString(msg_yang_operation_failed) + `</error-message>
	<error-info>
		<` + missing_choice_info.String() + ` xmlns="` + yang_namespace + `">` + html.EscapeString(name) + `</` + missing_choice_info.String() + `>
	</error-info>
</rpc-error>`
}

func TestMissingChoiceError(t *testing.T) {
	const (
		path = "/foo/bar"
		name = "baz"
	)
	ncerr := NewMissingChoiceError(path, name)
	marshal, err := json.MarshalIndent(ncerr, "", "\t")
	if err != nil {
		t.Errorf("Marshal MissingChoiceError error: %v\n", err)
		return
	}
	unmarshal := MissingChoiceError{}
	if err := json.Unmarshal(marshal, &unmarshal); err != nil {
		t.Errorf("Unmarshal MissingChoiceError error: %v\n", err)
		return
	}
	cmpMgmtError(t, ncerr.MgmtError, unmarshal.MgmtError)

	verifyXmlMarshal(t, ncerr, genMissingChoiceXml(path, name))
}

func genInsertFailedXml() string {
	return `<rpc-error xmlns="` + netconf_namespace + `">
	<error-type>` + html.EscapeString(error_type) + `</error-type>
	<error-tag>` + html.EscapeString(bad_attribute.String()) + `</error-tag>
	<error-severity>` + html.EscapeString(yang_severity_error.String()) + `</error-severity>
	<error-app-tag>` + html.EscapeString(missing_instance.String()) + `</error-app-tag>
	<error-message>` + html.EscapeString(msg_yang_bad_attribute) + `</error-message>
</rpc-error>`
}

func TestInsertFailedError(t *testing.T) {
	ncerr := NewInsertFailedError()
	marshal, err := json.MarshalIndent(ncerr, "", "\t")
	if err != nil {
		t.Errorf("Marshal InsertFailedError error: %v\n", err)
		return
	}
	unmarshal := InsertFailedError{}
	if err := json.Unmarshal(marshal, &unmarshal); err != nil {
		t.Errorf("Unmarshal InsertFailedError error: %v\n", err)
		return
	}
	cmpMgmtError(t, ncerr.MgmtError, unmarshal.MgmtError)

	verifyXmlMarshal(t, ncerr, genInsertFailedXml())
}


vyerror_test.go
// Copyright (c) 2017,2019, AT&T Intellectual Property. All rights reserved.
//
// Copyright (c) 2017 by Brocade Communications Systems, Inc.
// All rights reserved.
//
// SPDX-License-Identifier: MPL-2.0

package mgmterror

import (
	"fmt"
)

func ExampleExecError() {
	path := []string{"usr", "bin", "app"}
	err := NewExecError(path, "core dumped")
	fmt.Println(err.Error())

	//Output:
	// Error: /usr/bin/app: core dumped
}

func ExamplePathAmbiguousError() {
	path := []string{"s"}
	matches := map[string]string{
		"system":   "System parameters",
		"service":  "Services",
		"security": "Security",
	}
	err := NewPathAmbiguousError(path, matches)
	fmt.Println(err.Error())

	//Output:
	// Error: /s: Ambiguous path, could be one of: security, service, system
}

func ExamplePathAmbiguousError_command() {
	matches := map[string]string{
		"show": "Show the configuration (default values may be suppressed)",
		"set":  "Set the value of a parameter or create a new element",
		"save": "Save configuration to a file",
	}
	err := NewPathAmbiguousError([]string{}, matches)
	fmt.Println(err.Error())

	//Output:
	// Error: Ambiguous command, could be one of: save, set, show
}


doc.go
// Copyright (c) 2019, AT&T Intellectual Property. All rights reserved.
//
// Copyright (c) 2017 by Brocade Communications Systems, Inc.
// All rights reserved.
//
// SPDX-License-Identifier: MPL-2.0

/*

	Package mgmterror provides the standard NETCONF and YANG errors
	as defined in RFC6241 and RFC6020 respectively.

	While RFC6241 section 4.3 specifies 7 elements of an
	rpc-error, not all of them are mandatory for a specific
	error. For example, Path (error-path), will not be present if
	no appropriate payload element or datastore node can be
	associated with a particular error condition.

	Since the marshal and unmarshal interfaces require the
	MgmtError structure's members be public, they can be set as
	needed after an error's constructor is called.

	The specifications allow the Info (error-info) field to
	include additional elements to provide extended and/or
	implementation-specific debugging information. For example, an
	error from a low level library can add additional information
	about the error so that a higher layer can use that
	information.

*/
package mgmterror


vyerror.go
// Copyright (c) 2017-2019, AT&T Intellectual Property.  All rights reserved.
//
// Copyright (c) 2017 by Brocade Communications Systems, Inc.
// All rights reserved.
//
// SPDX-License-Identifier: MPL-2.0

package mgmterror

import (
	"bytes"
	"encoding/json"
	"encoding/xml"
	"fmt"
	"strings"

	"github.com/danos/utils/natsort"
	"github.com/danos/utils/pathutil"
)

type vyErrTag uint

const (
	vyatta_operation_failed vyErrTag = iota
)

var vyErrTagMap = map[string]vyErrTag{
	"operation-failed": vyatta_operation_failed,
}

func (t vyErrTag) String() string {
	for s, v := range vyErrTagMap {
		if t == v {
			return s
		}
	}
	// Can not happen
	return ""
}

type vyErrAppTagId uint

const (
	exec_failed vyErrAppTagId = iota
	path_ambig
)

var vyErrAppTagMap = map[string]vyErrAppTagId{
	"exec-failed":    exec_failed,
	"path-ambiguous": path_ambig,
}

func (t vyErrAppTagId) String() string {
	for s, v := range vyErrAppTagMap {
		if t == v {
			return s
		}
	}
	// Can not happen
	return ""
}

type vyAppTagMap map[vyErrAppTagId]interface{}

type vyError struct {
	severity yerrseverity
	msg      string
	apptag   vyAppTagMap
}

var vyErrTable map[vyErrTag]vyError

func init() {
	vyErrTable = map[vyErrTag]vyError{
		vyatta_operation_failed: {
			severity: yang_severity_error,
			msg:      msg_yang_operation_failed,
			apptag: vyAppTagMap{
				exec_failed: createExecError,
				path_ambig:  createPathAmbigError,
			},
		},
	}
}

func getVyattaError(err *MgmtError) error {
	tag, ok := vyErrTagMap[err.Tag]
	if !ok {
		return nil
	}
	errtag, ok := vyErrTable[tag]
	if !ok {
		return nil
	}
	apptag, ok := vyErrAppTagMap[err.AppTag]
	if !ok {
		return nil
	}
	fn, ok := errtag.apptag[apptag]
	if !ok {
		return nil
	}
	return callCreate(fn, err)
}

func (e *MgmtError) setVyattaError(tag vyErrTag, apptag, path string, info *MgmtErrorInfo) error {
	vyErr, ok := vyErrTable[tag]
	if !ok {
		return invalid_error_tag
	}
	e.Tag = tag.String()
	e.Typ = "application"
	e.Severity = vyErr.severity.String()
	e.Message = vyErr.msg
	e.AppTag = apptag
	e.Path = path
	if info != nil {
		e.Info = *info
	}
	return nil
}

func newVyattaError(tag vyErrTag, apptag, path string, info *MgmtErrorInfo) *MgmtError {
	e := newMgmtError()
	if err := e.setVyattaError(tag, apptag, path, info); err != nil {
		panic(err)
	}
	return e
}

type ExecError struct {
	*MgmtError
}

func (e *ExecError) UnmarshalJSON(value []byte) error {
	e.MgmtError = newMgmtError()
	return json.Unmarshal(value, e.MgmtError)
}

func (e *ExecError) MarshalXML(enc *xml.Encoder, start xml.StartElement) error {
	return enc.Encode(e.MgmtError)
}

func createExecError(err *MgmtError) *ExecError {
	return &ExecError{
		MgmtError: err,
	}
}

// A custom wrapper of a standard "operation failed" to represent an
// error when executing subtasks.
//
// path is the path of the subtask that was run
// out is the output of the subtask
func NewExecError(path []string, out string) *ExecError {
	err := newVyattaError(vyatta_operation_failed, exec_failed.String(),
		pathutil.Pathstr(path), nil)
	err.Message = out
	return createExecError(err)
}

type PathAmbiguousError struct {
	*MgmtError
}

func (e *PathAmbiguousError) UnmarshalJSON(value []byte) error {
	e.MgmtError = newMgmtError()
	return json.Unmarshal(value, e.MgmtError)
}

func (e *PathAmbiguousError) MarshalXML(enc *xml.Encoder, start xml.StartElement) error {
	return enc.Encode(e.MgmtError)
}

func (e *PathAmbiguousError) GetMessage() string {
	var b bytes.Buffer
	b.WriteString(ErrPath(pathutil.Makepath(e.Path)))
	b.WriteString(" is ambiguous\n")

	b.WriteString("Possible completions:\n")
	pathMap := make(map[string]string, len(e.Info))
	for _, elem := range e.Info {
		pathMap[elem.XMLName.Local] = elem.Value
	}
	sorted := make([]string, 0, len(pathMap))
	for n, _ := range pathMap {
		sorted = append(sorted, n)
	}

	natsort.Sort(sorted)
	for i, name := range sorted {
		b.WriteString(fmt.Sprintf("  %s\t%s", name, pathMap[name]))
		if i != len(sorted)-1 {
			b.WriteString("\n")
		}
	}
	return b.String()
}

func (e *PathAmbiguousError) Error() string {
	var mlist []string
	for _, t := range e.Info {
		mlist = append(mlist, t.XMLName.Local)
	}
	natsort.Sort(mlist)

	var b bytes.Buffer
	b.WriteString(strings.Title(e.Severity))
	b.WriteString(error_msg_separator)
	if len(e.Path) == 0 {
		b.WriteString("Ambiguous command")
	} else {
		b.WriteString(e.Path)
		b.WriteString(error_msg_separator)
		b.WriteString("Ambiguous path")
	}
	b.WriteString(", could be one of: ")
	for i, m := range mlist {
		if i > 0 {
			b.WriteString(", ")
		}
		b.WriteString(m)
	}
	return b.String()
}

func createPathAmbigError(err *MgmtError) *PathAmbiguousError {
	return &PathAmbiguousError{
		MgmtError: err,
	}
}

// A custom wrapper of a standard "operation failed" to represent an
// error when a path is ambiguous
//
// path is the path that was ambiguous
// matches are the possible completions for path
func NewPathAmbiguousError(path []string, matches map[string]string) *PathAmbiguousError {
	var info MgmtErrorInfo
	for k, v := range matches {
		t := NewMgmtErrorInfoTag(VyattaNamespace, k, v)
		info = append(info, *t)
	}
	err := newVyattaError(vyatta_operation_failed, path_ambig.String(),
		pathutil.Pathstr(path), &info)
	return createPathAmbigError(err)
}


mgmterror.go
// Copyright (c) 2017,2019-2020, AT&T Intellectual Property.
// All rights reserved.
//
// Copyright (c) 2016-2017 by Brocade Communications Systems, Inc.
// All rights reserved.
//
// SPDX-License-Identifier: MPL-2.0

package mgmterror

import (
	"bytes"
	"encoding/json"
	"encoding/xml"
	"errors"
	"fmt"
	"reflect"
	"strings"
)

const (
	// RFC6241 Sect 3.1
	netconf_module    = "ietf-netconf"
	netconf_namespace = "urn:ietf:params:xml:ns:netconf:base:1.0"

	// RFC6020 Sect 5.3.1
	yang_module    = "ietf-yang" // There is no yang module
	yang_namespace = "urn:ietf:params:xml:ns:yang:1"

	// TODO: need to find a better place for this
	vyattaModule    = "vyatta-yang"
	VyattaNamespace = "urn:vyatta.com:mgmt:error:1"

	// Used to separate fields in error message strings
	error_msg_separator = ": "
)

// based on RFC6241 Sect 4.3

type errtype uint

const (
	// Secure Transport
	transport errtype = iota

	// Messages
	rpc

	// Operations
	protocol

	// Content
	application
)

var errtypemap = map[string]errtype{
	"transport":   transport,
	"rpc":         rpc,
	"protocol":    protocol,
	"application": application,
}

func (t *errtype) set(typ string) error {
	if v, ok := errtypemap[typ]; ok {
		*t = v
		return nil
	}
	return errors.New("Invalid error target")
}

func (t errtype) String() string {
	for s, v := range errtypemap {
		if t == v {
			return s
		}
	}
	return ""
}

// MgmtErrorInfoTag holds additional information for an error.
//
// Some NETCONF and YANG defined errors have mandatory InfoTag
// information. Implementations may include additional elements to
// provide extended and/or implementation- specific debugging
// information.
type MgmtErrorInfoTag struct {
	XMLName xml.Name
	Value   string `xml:",chardata"`
}

// RFC7951 section 4 dictates that module name instead of namespace is
// used to differentiate elements when JSON encoded.
//
// Since we only have 3 namespaces to worry about right now, we can do
// the mapping locally. In the future, yangd can get a method to
// lookup namespaces/modules.
func (i *MgmtErrorInfoTag) lookupNamespace(module string) string {
	modmap := map[string]string{
		netconf_module: netconf_namespace,
		yang_module:    yang_namespace,
		vyattaModule:   VyattaNamespace,
	}
	ns, ok := modmap[module]
	if !ok {
		return module
	}
	return ns
}

func (i *MgmtErrorInfoTag) lookupModule(ns string) string {
	nsmap := map[string]string{
		netconf_namespace: netconf_module,
		yang_namespace:    yang_module,
		VyattaNamespace:   vyattaModule,
	}
	module, ok := nsmap[ns]
	if !ok {
		return ns
	}
	return module
}

func (i *MgmtErrorInfoTag) UnmarshalJSON(value []byte) error {
	var obj map[string]string
	if err := json.Unmarshal(value, &obj); err != nil {
		return err
	}
	if len(obj) != 1 {
		return errors.New("malformed error-info tag")
	}
	for k, v := range obj {
		s := strings.Split(k, ":")
		if len(s) == 2 {
			i.XMLName.Space = i.lookupNamespace(s[0])
			i.XMLName.Local = s[1]
		} else {
			i.XMLName.Local = s[0]
		}
		i.Value = v
	}
	return nil
}

func (i *MgmtErrorInfoTag) MarshalJSON() ([]byte, error) {
	var tag string
	var out bytes.Buffer
	out.WriteString("{")
	if len(i.XMLName.Space) > 0 {
		tag = i.lookupModule(i.XMLName.Space) + ":" + i.XMLName.Local
	} else {
		tag = i.XMLName.Local
	}
	b, err := json.Marshal(tag)
	if err != nil {
		return []byte(""), err
	}
	out.Write(b)
	out.WriteString(":")
	b, err = json.Marshal(i.Value)
	if err != nil {
		return []byte(""), err
	}
	out.Write(b)
	out.WriteString("}")
	return out.Bytes(), nil
}

func NewMgmtErrorInfoTag(ns, name, value string) *MgmtErrorInfoTag {
	return &MgmtErrorInfoTag{
		XMLName: xml.Name{
			Space: ns,
			Local: name,
		},
		Value: value,
	}
}

type MgmtErrorInfo []MgmtErrorInfoTag

func (e *MgmtErrorInfo) MarshalXML(enc *xml.Encoder, start xml.StartElement) error {
	error_info_name := xml.Name{
		Local: "error-info",
	}

	if err := enc.EncodeToken(xml.StartElement{Name: error_info_name}); err != nil {
		return err
	}
	for _, v := range *e {
		if err := enc.Encode(v); err != nil {
			return err
		}
	}
	if err := enc.EncodeToken(xml.EndElement{Name: error_info_name}); err != nil {
		return err
	}
	return nil
}

func (e *MgmtErrorInfo) UnmarshalXML(dec *xml.Decoder, start xml.StartElement) error {
	var infos []MgmtErrorInfoTag
Loop:
	for {
		tok, _ := dec.Token()
		if tok == nil {
			return nil
		}
		switch elem := tok.(type) {
		case xml.StartElement:
			var i MgmtErrorInfoTag
			if err := dec.DecodeElement(&i, &elem); err != nil {
				return err
			}
			infos = append(infos, i)
		case xml.EndElement:
			break Loop
		}
	}
	*e = infos
	return nil
}

func (e MgmtErrorInfo) FindMgmtErrorTag(ns, name string) string {
	for _, t := range e {
		if t.XMLName.Space == ns && t.XMLName.Local == name {
			return t.Value
		}
	}
	return ""
}

// Error structure based on RFC6241 Sect 4.3.
type MgmtError struct {
	XMLName xml.Name `json:"-"`

	// Typ is the type of error
	//
	// RFC6241 Sect 4.3 defines four types of errors:
	//  - transport (layer: Secure Transport)
	//  - rpc (layer: Messages)
	//  - protocol (layer: Operations)
	//  - application (layer: Content)
	Typ string `xml:"error-type" json:"error-type"`

	// Tag identifies the error condition
	//
	// Allowed values for NETCONF and YANG tags are defined by
	// RFC6241 Appendix A and RFC6020 Section 13 respectively.
	Tag string `xml:"error-tag" json:"error-tag"`

	// Severity indicates the error severity
	//
	// RFC6241 Sect 4.3 defines two of error severities:
	//  - error
	//  - warning
	Severity string `xml:"error-severity" json:"error-severity"`

	// AppTag identifies the data-model-specific or
	// implementation-specific error condition, if one exists.
	AppTag string `xml:"error-app-tag,omitempty" json:"error-app-tag,omitempty"`

	// Path contains the absolute XPath expression identifying the
	// element path to the node that is associated with the error
	// being reported if an appropriate payload element or
	// datastore node can be associated with a particular error
	// condition.
	//
	// See RFC6241 Section 4.3 for details on the context for the
	// XPath expression.
	Path string `xml:"error-path,omitempty" json:"error-path,omitempty"`

	// Message contains a string suitable for human display that
	// describes the error condition.
	Message string `xml:"error-message,omitempty" json:"error-message,omitempty"`

	// Info contains protocol- or data-model-specific error
	// content. This element will not be present if no such error
	// content is provided for a particular error condition. An
	// implementation MAY include additional elements to provide
	// extended and/or implementation- specific debugging
	// information.
	Info MgmtErrorInfo `xml:"error-info,omitempty" json:"error-info,omitempty"`
}

func newMgmtError() *MgmtError {
	e := &MgmtError{}
	e.setXMLName()
	return e
}

// MgmtErrorRef - interface that allows us to identify all types of MgmtError
// in a single check.  Use of private function (mgmtErrorRef) ensures no one
// else can create an object that meets the interface unless it explicitly
// includes a MgmtError object, in which case that's ok.
type MgmtErrorRef interface {
	mgmtErrorRef()
}

var _ MgmtErrorRef = (*MgmtError)(nil)

func (me *MgmtError) mgmtErrorRef() {}

// Formattable - interface provided by error types to allow formatting
// (NB: MgmtError is just one example of such a type.)
//
// Unlike plain Go errors, MgmtError types have useful properties which
// we may wish to print out differently depending on the end user / context.
// GetMessage() may well be customised for certain error types to include
// the contents of MgmtError.Info.
//
type Formattable interface {
	GetMessage() string
	GetPath() string
	GetSeverity() string
	GetTag() string
	GetAppTag() string
	GetType() string
	GetInfo() MgmtErrorInfo
}

// Ensure *MgmtError implements interface
var _ Formattable = (*MgmtError)(nil)

func (me *MgmtError) GetMessage() string     { return me.Message }
func (me *MgmtError) GetPath() string        { return me.Path }
func (me *MgmtError) GetSeverity() string    { return me.Severity }
func (me *MgmtError) GetTag() string         { return me.Tag }
func (me *MgmtError) GetAppTag() string      { return me.AppTag }
func (me *MgmtError) GetType() string        { return me.Typ }
func (me *MgmtError) GetInfo() MgmtErrorInfo { return me.Info }

func callCreate(fn interface{}, err *MgmtError) error {
	ty := reflect.TypeOf(fn)
	if ty.Kind() != reflect.Func ||
		ty.NumIn() != 1 ||
		ty.NumOut() != 1 {
		return err
	}
	vfn := reflect.ValueOf(fn)
	result := vfn.Call([]reflect.Value{reflect.ValueOf(err)})
	if len(result) != 1 {
		return err
	}
	e, ok := result[0].Interface().(error)
	if !ok {
		return err
	}
	return e
}

func (e *MgmtError) setXMLName() {
	e.XMLName = xml.Name{
		Space: netconf_namespace,
		Local: "rpc-error",
	}
}

func (e MgmtError) Error() string {
	var b bytes.Buffer

	b.WriteString(strings.Title(e.Severity))
	b.WriteString(error_msg_separator)

	if e.Path != "" {
		b.WriteString(e.Path)
		b.WriteString(error_msg_separator)
	}

	if e.Message != "" {
		b.WriteString(e.Message)
	}

	return b.String()
}

const errpfx = "com.vyatta.rpcerror."

// Encode error for DBus
func (e *MgmtError) DBusError() (string, []interface{}) {
	name := fmt.Sprintf("%s%s", errpfx, e.Typ)
	body := make([]interface{}, 1)
	body[0] = e
	return name, body
}
